## Detailed iSTAR manual 

---

### 0. Introduction

Github of iSTAR: https://github.com/daviddaiweizhang/istar

Paper: https://doi.org/10.1038/s41587-023-02019-9

### 1. Input files

It requirs 5 files as input.

![](https://github.com/Liying1996/iSTAR_Manual/blob/main/figs/data_format.png)

#### 1.1 he-raw.jpg

Ensure that the histology image is the **raw** image, not a high-resolution version.

#### 1.2 cnts.tsv

This is the cell-gene count matrix.

- Column headers should be gene names, with rows (starting from the second row) representing gene expression levels.
- Row names are spot, which should match exactly with those in `locs-raw.tsv` in terms of names and order. Duplicates are not allowed.

Example:

![](https://github.com/Liying1996/iSTAR_Manual/blob/main/figs/cnts-example.png)

#### 1.3 locs-raw.tsv

This file contains position information that can be extracted from `tissue_positions.csv` or `tissue_positions_list.csv`, generated by Cell Ranger in the `spatial` folder.

- Only the 1st, 5th, and 6th columns are needed.
  - `pxl_row_in_fullres` corresponds to **y** and `pxl_col_in_fullres` corresponds to **x**.
- The row names and order must match exactly with those in `cnts.tsv`.

Example:

![](https://github.com/Liying1996/iSTAR_Manual/blob/main/figs/locs-raw-example.png)

#### 1.4 pixel-size-raw.txt

For [Visium](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/spatial) data, this value can be approximated by `8000 / 2000 * tissue_hires_scalef`, where `tissue_hires_scalef` is stored in `scalefactors_json.json`

Example:

![](https://github.com/Liying1996/iSTAR_Manual/blob/main/figs/scalefactors_json.png)



Then the pixel-size-raw is:

8000 / 2000 * tissue_hires_scalef  = 0.64

#### 1.5 radius-raw.txt

For [Visium](https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/spatial) data, this value can be computed by `spot_diameter_fullres * 0.5`, where `spot_diameter_fullres` is stored in `scalefactors_json.json`, and should be close to `55 * 0.5 / pixel_size_raw`.

---

## 2. Running the code

Install the required packages:

```
pip install -r requirements.txt
```

Run the code as follows:

```
sh run.sh the_folder_includes_allinputfiles
```

The code :https://github.com/daviddaiweizhang/istar/blob/master/run.sh

![](https://github.com/Liying1996/iSTAR_Manual/blob/main/figs/run-code.png)

**Note**: If a different rescaled pixel size or top *n* genes is required, adjust these values directly in the original code.

---

## 3. Step-by-Step Guide

#### 3.1 Install

```
!git clone https://github.com/daviddaiweizhang/istar.git
!pip install -r istar/requirements.txt
!./download_checkpoints.sh
```

#### 3.2 Preprocess histology image

```
# !echo 0.6 > input_iSTAR/input/pixel-size-raw.txt 
!echo 0.5 > input_iSTAR/input/pixel-size.txt
!python rescale.py input_iSTAR/input/ --image
!python preprocess.py input_iSTAR/input/ --image
```

A higher (rescaled) pixel size (second row) leads to higher resolution for clustering, though it also requires more processing time.

#### 3.3 Extract features and get mask of image

```
!python extract_features.py input_iSTAR/input/ --device="cuda"
!python get_mask.py input_iSTAR/input/embeddings-hist.pickle input_iSTAR/input/mask-small.png
```

#### 3.4 Rescale coordinates and spot radius of image

```
!python rescale.py input_iSTAR/input/ --locs --radius
```

#### 3.5 Select top genes

```
!python select_genes.py --n-top=1000 "input_iSTAR/input/cnts.tsv" "input_iSTAR/input/gene-names.txt"
```

The default is the top 1,000 genes. If you wish to use a custom gene list, create a file named `gene-names.txt`.

#### 3.6 Train, predict and visulize at super-resolution

```
!python impute.py input_iSTAR/input/ --epochs=500 --device="cuda"  # train model from scratch
!python plot_imputed.py "input_iSTAR/input/"
# If you want to retrain model, you need to delete the existing model:
# rm -r ${prefix}states
```

#### 3.7 Segment image by gene features

```
!python cluster.py --filter-size=8 --n-clusters=40 --mask=ETC_B2S2/input/mask-small.png ETC_B2S2/input/embeddings-gene.pickle ETC_B2S2/input/clusters-gene/
```

---

#### Additional

iSTAR defaults to using min-max normalization. If this is not needed (e.g., if you have already performed normalization using CP10K or similar methods and want to compare signal intensities across different samples or time points), you can comment out the relevant code in `impute.py` and `plot_impute.py`. For example, as shown below.

(impute.py)

[](https://github.com/Liying1996/iSTAR_Manual/blob/main/figs/impute.png)

```
def normalize(embs, cnts):

    embs = embs.copy()
    cnts = cnts.copy()

    # TODO: check if adjsut_weights in extract_features can be skipped
    embs_mean = np.nanmean(embs, (0, 1))
    embs_std = np.nanstd(embs, (0, 1))
#    embs -= embs_mean
#    embs /= embs_std + 1e-12

    cnts_min = cnts.min(0)
    cnts_max = cnts.max(0)
#    cnts -= cnts_min
#    cnts /= (cnts_max - cnts_min) + 1e-12

    return embs, cnts, (embs_mean, embs_std), (cnts_min, cnts_max)
```

(plot_impute.py)

[](https://github.com/Liying1996/iSTAR_Manual/blob/main/figs/plot_impute.png)

-> comment out normalization part and change the color mapping

```
def plot_super(
    x, outfile, underground=None, truncate=None):

	x = x.copy()
	mask = np.isfinite(x)

	if truncate is not None:
	    x -= np.nanmean(x)
	    x /= np.nanstd(x) + 1e-12
	    x = np.clip(x, truncate[0], truncate[1])

	cmap = plt.get_cmap('turbo')

	if underground is not None:
	    under = underground.mean(-1, keepdims=True)
	    under -= under.min()
	    under /= under.max() + 1e-12

	img = cmap(np.clip(x, 0, 0.02)/0.02)[..., :3] # or change 0.02 to other max number you need

	if underground is not None:
	    img = img * 0.5 + under * 0.5

	img[~mask] = 1.0

	img = (img * 255).astype(np.uint8)
```

